File: 46.rb
# https://leetcode.com/problems/permutations/
# @param {Integer[]} nums
# @return {Integer[][]}
def permute(nums, tmp=[], res=[])
	return [] if !nums || nums.empty?

	if tmp.size == nums.size
		res << tmp.dup
		return
	end

	nums.each { |num|
		if !tmp.include?(num)
			tmp << num
			permute(nums, tmp, res)
			tmp.pop
		end
	}
	res
end

File: js/46.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums, temp=[], res=[]) {
	if (!nums || nums.length === 0) {
		return [[]]
	}

	if (nums.length === temp.length) {
		res.push([...temp])
		temp = []
		return
	}

	nums.forEach((el) => {
		if (temp.indexOf(el) === -1) {
			temp.push(el)
			permute(nums, temp, res)
			temp.pop()
		}
	})
	return res
};
